File: aidlclient/src/main/java/com/yikwing/aidlclient/MainActivity.java
package com.yikwing.aidlclient;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.Bundle;
import android.os.IBinder;
import android.os.RemoteException;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.widget.Button;
import com.yikwing.aidlserver.IMyAidlInterface;

/**
 * @author yikwing
 * @date 2019/2/20 4:29 PM
 * 温馨提示
 * 代码千万行
 * 注释第一行
 * 命名不规范
 * 同事两行泪
 */
public class MainActivity extends AppCompatActivity {


    private IMyAidlInterface mIMyAidlInterface;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        Button btn = findViewById(R.id.btnCLick);


        Intent intent = new Intent();
        // 服务端AndroidManifest.xml文件该Service所配置的action
        intent.setAction("com.yikwing.aidlserver");
        // Service所在的包名
        intent.setPackage("com.yikwing.aidlserver");
        bindService(intent, ConnectCallBack, Context.BIND_AUTO_CREATE);


        btn.setOnClickListener(v -> {


            if (mIMyAidlInterface != null) {
                try {
                    mIMyAidlInterface.login("yikwing", "123456");
                } catch (RemoteException e) {
                    e.printStackTrace();
                }
            }

        });

    }


    private ServiceConnection ConnectCallBack = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {
            mIMyAidlInterface = IMyAidlInterface.Stub.asInterface(service);
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {
            mIMyAidlInterface = null;
        }
    };


//    class ConnectCallBack implements ServiceConnection {
//
//        @Override
//        public void onServiceConnected(ComponentName name, IBinder service) {
//            mIMyAidlInterface = IMyAidlInterface.Stub.asInterface(service);
//        }
//
//        @Override
//        public void onServiceDisconnected(ComponentName name) {
//            mIMyAidlInterface = null;
//        }
//    }
}

File: app/src/main/java/com/yikwing/aidldemo/MainActivity.kt
package com.yikwing.aidldemo

import android.content.ComponentName
import android.content.Intent
import android.content.ServiceConnection
import android.os.Bundle
import android.os.IBinder
import android.os.RemoteException
import android.support.v7.app.AppCompatActivity
import android.widget.Toast
import com.yikwing.aidldemo.aidl.Book
import com.yikwing.aidldemo.aidl.IBookManager
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {

    var mIBookManager: IBookManager? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        onBinder.setOnClickListener {
            val intent = Intent()
            intent.action = "com.yikwing.aidldemo.aidl.BookManagerService"
            intent.run {
                setPackage(packageName)
                setFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            }
            bindService(intent, mServiceConnection, BIND_AUTO_CREATE)
        }

        onAdd.setOnClickListener {
            if (mIBookManager != null) {
                try {
                    mIBookManager?.addBook(Book(18, "漫画书"));
                } catch (e: RemoteException) {
                    e.printStackTrace()
                }
            }
        }

    }


    private val mServiceConnection = object : ServiceConnection {
        override fun onServiceDisconnected(name: ComponentName?) {
            mIBookManager = null
        }

        override fun onServiceConnected(name: ComponentName?, service: IBinder?) {
            mIBookManager = IBookManager.Stub.asInterface(service)
            Toast.makeText(this@MainActivity, "绑定成功", Toast.LENGTH_SHORT).show()
        }

    }


}
